sysboundary/pml.cpp

```diff
@@ -0,0 +1,275 @@
+/*
+ * This file is part of Vlasiator.
+ * Copyright 2010-2016 Finnish Meteorological Institute
+ *
+ * For details of usage, see the COPYING file and read the "Rules of the Road"
+ * at http://www.physics.helsinki.fi/vlasiator/
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+/*!\file mpl.cpp
+ * \builds the arrays needed for the PML.
+ */
+
+#include "pml.h"
+#include <cstdlib>
+#include <iostream>
+#include <iomanip> 
+#include <cmath>
+#include <sstream>
+#include <ctime>
+#include <omp.h>
+#include "fsgrid.hpp"
+#include "../definitions.h"
+#include "../common.h"
+#include "../parameters.h"
+#include "../readparameters.h"
+
+
+
+
+bool buildPMLGrids(
+    FsGrid<std::array<Real, fsgrids::pml::N_PML>, 2> &pmlGrid)
+{
+
+   typedef Parameters P;
+
+
+   /*----Get PML Array and Domain Size-----*/
+   std::array<Real, fsgrids::pml::N_PML> *pmlValue;
+   int start = 2;   /*don't judge will change it */
+   std::array<int32_t, 3> pos;
+   const int *pmlDims = &pmlGrid.getLocalSize()[0];
+   const int *globalDims = &pmlGrid.getGlobalSize()[0];
+
+
+
+   /*Read PLM faces to build*/
+   std::vector<std::string> thisSpeciesFaceList;
+   Readparameters::get("PML.face", thisSpeciesFaceList) ;
+   bool Xp, Xm, Yp, Ym, Zp, Zm = false;
+   for (auto &face : thisSpeciesFaceList){
+   if(face == "x+") {Xp = true;}      //set face to true  
+   if(face == "x-") {Xm = true;}      //set face to true
+   if(face == "y+") {Yp = true;}      //set face to true
+   if(face == "y-") {Ym = true;}      //set face to true
+   if(face == "z+") {Zp = true;}      //set face to true
+   if(face == "z-") {Zm = true;}      //set face to true  
+   }   
+
+
+    /*-----Initially set all arrays to one-----*/
+    /*Iterate over domain and set the PML arrays to 1.0 thus not affecting the fieldsolver*/ 
+    #pragma omp parallel for collapse(3)
+    for (int kk = 0; kk < pmlDims[2]; kk++){
+        for (int jj = 0; jj < pmlDims[1]; jj++){
+            for (int ii = 0; ii < pmlDims[0]; ii++){
+
+                pmlValue = pmlGrid.get(ii, jj, kk);
+                pmlValue->at(fsgrids::pml::PGI2) = 1.0;
```

Comment:
Since `fsgrids::pml` is an enum, you could have avoided typing pain and done a loop from 0 to N_PML. This would actually reduce the maintenance issues if this array changes later.